Simplify logging

antkmsft opened the pull request:
Closes #1180.

JeffreyRichter commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -211,13 +230,21 @@ CURLcode CurlSession::Perform(Context const& context)
     if (isContentLengthHeaderInRequest == headers.end())
     {
-      LogThis("No content-length in headers. Adding it");
+      if (ShouldWrite(LogLevel::Verbose))
```
You can leave this pattern as is, but since these strings are concatenated at compile time, you could NOT call ShouldLog first and just call Log.

JeffreyRichter commented on sdk/core/azure-core/inc/azure/core/internal/log.hpp:
```diff
@@ -5,7 +5,7 @@
-  void Write(LogClassification const& classification, std::string const& message);
-}}}} // namespace Azure::Core::Logging::Details
+namespace Azure { namespace Core { namespace Logging { namespace Internal {
+  bool ShouldWrite(LogLevel level);
```
Can we call these ShouldLog and Log? This makes it clear when looking at code when logging is happening.
Also, they take a LogLevel, not a WriteLevel.

JeffreyRichter commented on sdk/core/azure-core/src/logging/logging.cpp:
```diff
@@ -7,72 +7,41 @@
 namespace {
 std::mutex g_loggerMutex;
 LogListener g_logListener(nullptr);
+LogLevel g_logLevel = LogLevel::Verbose;
```
The default is either Error or maybe we need a None.
Also, this variable should be atomic.

antkmsft replied:
We need to sync both level and callback, std::function is not atomic. Other languages don't have None. If you want none, why have callback set?

JeffreyRichter replied:
If both variables are atomic, then you don't need the mutex.
The customer needs to provide a callback so they can write to the console, file, etc.
The callback can be set and then the level can be adjusted dynamically over the lifetime of the app.

antkmsft replied:
`std::atomic<std::function<...>>` does not compile, underlying type does not correspond to atomic requirements.

JeffreyRichter replied:
Could make it an atomic void* and then cast.

antkmsft replied:
I thought about that, I am not sure if std::function's pointer can always be represented as void*, or maybe under some circumstances it could have more. Because I'd expect `std::atomic<>` to also compile if the type is compatible, but apparently it's not. I'll talk to Casey.If we want this checked in today, let's defer this decision because I am not sure it is always safe to cast. It won't be a breaking change to update it in the future, it is an implementation detail, only visible there.

antkmsft replied:
@CaseyCarter, what are our options?

JeffreyRichter replied:
I agree Anton - can be changed later.

CaseyCarter replied:
`std::function` can represent function objects with state, for which there is no corresponding equivalent pointer-to-function. `std::atomic` requires its parameter to be trivially copyable - effectively a bag of bits that can be `memcpy`d - which `std::function` is not (if there's more state than will fit in the small object buffer inside the `std::function` it has to allocate memory).

Without having scrutinized the code (or looked at anything outside of this comment) why do we hate `mutex`?

JeffreyRichter replied:
I wouldn't say we "hate mutex". Keeping a mutex is fine. It's just that atomic is usually faster - but this is unlikely to be a performance-sensitive piece of code..

JeffreyRichter commented on sdk/core/azure-core/src/logging/logging.cpp:
```diff
@@ -7,72 +7,41 @@
-    LogClassificationsPrivate::LogClassificationsConstant(true));
-
-LogListener GetLogListener(LogClassification const& classification)
+LogListener GetLogListener(LogLevel level)
```
We don't expose this to customers, right?

antkmsft replied:
No, this is internal even to this only cpp file.

JeffreyRichter commented on sdk/core/azure-core/src/logging/logging.cpp:
```diff
@@ -7,72 +7,41 @@
+using namespace Azure::Core::Logging::Internal;
 
 namespace {
 std::mutex g_loggerMutex;
```
I think you could kill the mutex and just make the callback function atomic. This would make the code smaller and faster.

antkmsft replied:
`std::function` is not atomic, and also we want to lock both the level and the callback, or otherwise you may get unexpected levels passed to your your old/new callback.

JeffreyRichter replied:
We generally don't are about unexpected levels being passed to the callback - the callback HAS to be written to expect any level at any time.

antkmsft replied:
Unfortunately, `std::atomic<std::function<...>>` does not compile, underlying type does not correspond to atomic requirements.

JeffreyRichter commented on sdk/core/azure-core/src/logging/logging.cpp:
```diff
@@ -7,72 +7,41 @@
-      ? g_logListener // return actual listener (may be null)
-      : LogListener(nullptr) // return null listener
-      ;
+  return (g_logListener && g_logLevel <= level) ? g_logListener : LogListener(nullptr);
```
If you use atomic, you'd copy g_log_listener into a local variable and use the local variable twice on the return line.

JeffreyRichter commented on sdk/core/azure-core/inc/azure/core/logging/logging.hpp:
```diff
@@ -9,143 +9,51 @@
-      constexpr static auto Classification(int16_t number) { return LogClassification(F, number); }
-    };
-  } // namespace Details
+  void SetMaximumLogLevel(LogLevel level);
```
Im' not sure that "maximum" is a good term here. 
Maybe just SetLogLevel and the docs explains what it means.

RickWinter commented on sdk/core/azure-core/inc/azure/core/logging/logging.hpp:
```diff
@@ -9,143 +9,51 @@
   /**
-   * @brief Allows the application to specify which log classification types it is interested in
-   * receiving.
+   * @brief Sets the maximul #LogLevel an application is interested in receiving.
```
```suggestion
   * @brief Sets the #LogLevel an application is interested in receiving.
```

RickWinter commented on sdk/core/azure-core/inc/azure/core/logging/logging.hpp:
```diff
@@ -9,143 +9,51 @@
-   * @brief Allows the application to specify which log classification types it is interested in
-   * receiving.
+   * @brief Sets the maximul #LogLevel an application is interested in receiving.
+   * @detail Default is #LogLevel::Verbose.
```
```suggestion
   * @detail Default is #LogLevel::Informational.
```

Default should be informational

RickWinter replied:
Or default should be none.

antkmsft replied:
it is "None" in a sense that you get none until you set callback. If you set it, you are probably interested in at least errors?
I set Verbose, because I would expect to get, simply, all, when I set logger. Then if I get too much, I tune it down, but not vice versa - I may know about the `SetLogListener()` but not about `SetLogLevel()`. Then I don't get enough log messages, so I think the SDK is not instrumented well enough with log messages.

antkmsft replied:
Also, other languages don't have `LogLevel::None`

antkmsft replied:
To give us some room, I am simply leaving the default undocumented now, without changing the actual default value. I still think Verbose is the best default choice, in terms of user-friendliness.

RickWinter commented on sdk/core/azure-core/inc/azure/core/logging/logging.hpp:
```diff
@@ -9,143 +9,51 @@
+  /**
+   * @brief Log message level.
+   */
+  enum class LogLevel
```
If these were flags then you could pick which type you want and it isn't stacking.
Error = 1,
Warning = 2, 
Informational = 4,
Verbose = 8,
...

Setting the log level to 0 would be == no logging

antkmsft replied:
If you want no logging, you set listener to `nullptr`.
I don't think we need flags - who would want verbose and warnings, but not informational and errors? I think, it should be "maximum level", and that's it. If you want no logging, you unset the logger, what's the point to have logger that will get no messages?

JeffreyRichter replied:
You have to think about different concerns here. For example, it is common that an app will set one callback for the lifetime of the app but the level will change dynamically over the app's lifetime. To turn off logging, setting a null callback would work but then that code would have to know the name of the callback function to turn it back on again. Whereas the legal set of levels (error, warning, etc.) is always known globally throughout the program.

antkmsft replied:
I think of a LogLevel more as sound volume level, and not as bit flags.

JeffreyRichter replied:
I agree

antkmsft replied:
But let's say if you really want Warnings and Informational but not errors and Verbose for some reason, you can filter in your callback.

ahsonkhan replied:
What if I want all logging, or as @RickWinter mentioned both error and warning?

The bit flag approach makes the most sense to achieve that. If I pick Warning, am I automatically getting both Warning and Error?

JeffreyRichter replied:
Yes, this is how it works in most (maybe all) other languages. Logging is s scale; Verbose gives you everything.
Each log level is a severity and one severity implies the higher severities. It's not useful to ask for warnings and not ask for errors.

RickWinter commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -25,17 +25,7 @@
-        "[CURL Transport Adapter]: " + msg);
-  }
-}
+std::string const LogMsgPrefix = "[CURL Transport Adapter]: ";
```
This seems fine for now.  I suggest we standardize log message formats in the design guidelines

antkmsft replied:
I like the `string serviceId`. But let's discuss.

ahsonkhan replied:
curl isn't a service. How about `componentId`?

antkmsft replied:
I agree

vhvb1989 commented:
/azp run cpp - storage

antkmsft commented:
I addressed all the feedback. @RickWinter, @ahsonkhan, @vhvb1989, the issue this PR is targeting was selected for Tuesday release, so I need an approval today.

ahsonkhan reviewed (changes requested):
Temporarily blocking to avoid accidental merge since CI legs aren't running on this PR for some reason.

vhvb1989 commented:
> Temporarily blocking to avoid accidental merge since CI legs aren't running on this PR for some reason.

CI is running:  See https://dev.azure.com/azure-sdk/public/_build/results?buildId=715740&view=results

Looks like the connection to github checks is not working

ahsonkhan commented on sdk/core/azure-core/inc/azure/core/logging/logging.hpp:
```diff
@@ -9,143 +9,50 @@
    */
-  typedef std::function<void(LogClassification const& classification, std::string const& message)>
-      LogListener;
+  typedef std::function<void(LogLevel level, std::string const& message)> LogListener;
```
I assumed we would only call the user-defined `LogListener` if the LogLevel matches? Are we leaving the filtering to the caller?

Should this be:
```suggestion
  typedef std::function<void(std::string const& message)> LogListener;
```

antkmsft replied:
I don't think that is what we discussed, but it is an interesting idea. I'd it rather had the log level. We can change later/add an overload. If you think we should remove it, let's talk with the team.

vhvb1989 commented:
still failing: https://dev.azure.com/azure-sdk/public/_build/results?buildId=715999&view=results

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -25,17 +25,7 @@
 
 namespace {
-// Can be used from anywhere a little simpler
-inline void LogThis(std::string const& msg)
```
nit: keep the helper which appends `LogMsgPrefix` centrally, to avoid having to repeat appending it everywhere.

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -152,15 +150,18 @@ using Azure::Core::Http::CurlSession;
+
   // Create CurlSession to perform request
-  LogThis("Creating a new session.");
+  Log(LogLevel::Verbose, LogMsgPrefix + "Creating a new session.");
```
```suggestion
  Log(LogLevel::Informational, LogMsgPrefix + "Creating a new session.");
```

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -191,7 +192,9 @@ std::unique_ptr<RawResponse> CurlTransport::Send(Context const& context, Request
         "Error while sending request. " + std::string(curl_easy_strerror(performing)));
```
@vhvb1989 - let's review the places where we throw exceptions and log and `Error` level message following:
https://azure.github.io/azure-sdk/general_implementation.html#logging

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -226,21 +231,22 @@ CURLcode CurlSession::Perform(Context const& context)
   // somehow lost, libcurl will return CURLE_UNSUPPORTED_PROTOCOL
   // (https://curl.haxx.se/libcurl/c/curl_easy_send.html). Return the error back.
-  LogThis("Send request without payload");
+  Log(LogLevel::Verbose, LogMsgPrefix + "Send request without payload");
```
```suggestion
  Log(LogLevel::Informational, LogMsgPrefix + "Send request without payload");
```

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -226,21 +231,22 @@ CURLcode CurlSession::Perform(Context const& context)
   }
 
-  LogThis("Parse server response");
+  Log(LogLevel::Verbose, LogMsgPrefix + "Parse server response");
```
```suggestion
  Log(LogLevel::Informational, LogMsgPrefix + "Parse server response");
```

ahsonkhan commented on sdk/core/azure-core/src/http/curl/curl.cpp:
```diff
@@ -251,18 +257,17 @@ CURLcode CurlSession::Perform(Context const& context)
   if (this->m_lastStatusCode != HttpStatusCode::Continue)
   {
-    LogThis("Server rejected the upload request");
+    Log(LogLevel::Verbose, LogMsgPrefix + "Server rejected the upload request");
```
```suggestion
    Log(LogLevel::Informational, LogMsgPrefix + "Server rejected the upload request");
```

ahsonkhan commented on sdk/core/azure-core/src/http/logging_policy.cpp:
```diff
@@ -77,12 +77,11 @@ std::unique_ptr<RawResponse> Azure::Core::Http::LoggingPolicy::Send(
   }
-
-  if (!Logging::Details::ShouldWrite(LogClassification::Response))
+  else
```
Is this really the logic we want?

ahsonkhan replied:
From offline discussion, yes it is. We are still calling the next policy outside the else block, below.

ahsonkhan commented on sdk/core/azure-core/src/http/logging_policy.cpp:
```diff
@@ -91,8 +90,8 @@ std::unique_ptr<RawResponse> Azure::Core::Http::LoggingPolicy::Send(
 
-  Logging::Details::Write(
-      LogClassification::Response, GetResponseLogMessage(request, *response, end - start));
+  Logging::Internal::Log(
+      Logging::LogLevel::Verbose, GetResponseLogMessage(request, *response, end - start));
```
Shouldn't all these have a `ShouldLog` check?

ahsonkhan replied:
From offline discussion, no. From above `ShouldLog` is already returning true.

ahsonkhan commented on sdk/core/azure-core/src/logging/logging.cpp:
```diff
@@ -7,72 +7,41 @@
+bool Azure::Core::Logging::Internal::ShouldLog(LogLevel level)
 {
-  return GetLogListener(classification) != nullptr;
+  return GetLogListener(level) != nullptr;
```
Are we OK with getting a lock for every `shouldlog` check?

ahsonkhan commented on sdk/core/azure-core/test/ut/logging.cpp:
```diff
@@ -1,112 +1,191 @@
-  LogArguments Actual;
+  EXPECT_FALSE(ShouldLog(LogLevel::Verbose));
+  EXPECT_FALSE(ShouldLog(LogLevel::Informational));
+  EXPECT_FALSE(ShouldLog(LogLevel::Warning));
+  EXPECT_FALSE(ShouldLog(LogLevel::Error));
```
Default should be Error or Warning, nothing lower.

antkmsft commented:
Let's take more time to discuss this changes, we are not targeting Tuesday now. Put your ideas.

ahsonkhan reviewed (approved):
Looks good. Thanks!

ahsonkhan commented on sdk/core/azure-core/test/ut/logging.cpp:
```diff
@@ -1,112 +1,225 @@
+    }
+
+    SetLogLevel(LogLevel::Verbose);
+    {
+      level = LogLevel::Error;
+      message = "";
+
+      Log(LogLevel::Verbose, "Verbose");
+      EXPECT_EQ(level, LogLevel::Verbose);
+      EXPECT_EQ(message, "Verbose");
+
+      Log(LogLevel::Informational, "Informational");
+      EXPECT_EQ(level, LogLevel::Informational);
+      EXPECT_EQ(message, "Informational");
 
-  for (auto value : Values)
+      Log(LogLevel::Warning, "Warning");
+      EXPECT_EQ(level, LogLevel::Warning);
+      EXPECT_EQ(message, "Warning");
+
+      Log(LogLevel::Error, "Error");
+      EXPECT_EQ(level, LogLevel::Error);
+      EXPECT_EQ(message, "Error");
+    }
```
This block is duplicated above, remove?
We have four levels, so I expected four such blocks and we have five.

antkmsft replied:
No, we verify that you can set it back to verbose. Maybe it does not make as much sense, but that was intentional.

antkmsft commented on sdk/core/azure-core/test/ut/logging.cpp:
```diff
@@ -1,112 +1,225 @@
+      EXPECT_EQ(message, "Error");
+    }
+
+    SetLogLevel(LogLevel::Verbose);
```
```suggestion
    // Verify that we can switch back to Verbose 
    SetLogLevel(LogLevel::Verbose);
```
